diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 027a721..7eac3ff 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -23,8 +23,12 @@ else()
 set(SOURCES ${SOURCES} seshat_interface_stub.c)
 endif (ENABLE_SESHAT)
 
-
+if (PARODUS_SERVICE_APP)
 add_executable(parodus ${SOURCES})
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DPARODUS_SERVICE_APP ")
+else ()
+add_library(parodus SHARED ${SOURCES})
+endif (PARODUS_SERVICE_APP)
 
 target_link_libraries (parodus
     ${CMAKE_THREAD_LIBS_INIT}
@@ -52,5 +56,10 @@ if (ENABLE_SESHAT)
 target_link_libraries (parodus -llibseshat)
 endif (ENABLE_SESHAT)
 
+if (PARODUS_SERVICE_APP)
+set (INSTALL_DIR bin)
+else ()
+set (INSTALL_DIR lib)
+endif ()
 
-install (TARGETS parodus DESTINATION bin)
+install (TARGETS parodus DESTINATION ${INSTALL_DIR})
diff --git a/src/main.c b/src/main.c
index 37a63d4..04c0a20 100644
--- a/src/main.c
+++ b/src/main.c
@@ -51,23 +51,27 @@ static void sig_handler(int sig);
 /*----------------------------------------------------------------------------*/
 /*                             External Functions                             */
 /*----------------------------------------------------------------------------*/
-int main( int argc, char **argv)
+#ifdef PARODUS_SERVICE_APP
+int main( int argc, char **argv )
+#else
+int ParodusServiceMain( int argc, char **argv )
+#endif
 {
 #ifdef INCLUDE_BREAKPAD
     breakpad_ExceptionHandler();
 #else
     signal(SIGTERM, sig_handler);
-	signal(SIGINT, sig_handler);
-	signal(SIGUSR1, sig_handler);
-	signal(SIGUSR2, sig_handler);
-	signal(SIGSEGV, sig_handler);
-	signal(SIGBUS, sig_handler);
-	signal(SIGKILL, sig_handler);
-	signal(SIGFPE, sig_handler);
-	signal(SIGILL, sig_handler);
-	signal(SIGQUIT, sig_handler);
-	signal(SIGHUP, sig_handler);
-	signal(SIGALRM, sig_handler);
+    signal(SIGINT, sig_handler);
+    signal(SIGUSR1, sig_handler);
+    signal(SIGUSR2, sig_handler);
+    signal(SIGSEGV, sig_handler);
+    signal(SIGBUS, sig_handler);
+    signal(SIGKILL, sig_handler);
+    signal(SIGFPE, sig_handler);
+    signal(SIGILL, sig_handler);
+    signal(SIGQUIT, sig_handler);
+    signal(SIGHUP, sig_handler);
+    signal(SIGALRM, sig_handler);
 #endif	
     ParodusCfg *cfg;
 
